### DISCUSSION TITLE
How to type an object as being `Values` that can be passed in to a serializer or set state action?

### DISCUSSION THREAD

**Discussion by neefrehman** [47ng/nuqs#653] [Q&A] [answered]
I'm currently working on an abstracted pagination state hook, and would like to return an object that callers can choose to pass into a `serializePaginationSearchParams` or `setPaginationState` function (that the hook also returns). This would allow for simple clearing of state with a good level of composability (enabling callers to choose whether to "push" or "replace" state, for example). But, I'm struggling to type this object in a way that will be accepted by both functions, does anyone know how I could do this?

```ts
// simplified version of hook implementation

const [paginationState, setPaginationState] = useQueryStates(searchParams, {
    history: 'push',
});

const serializePaginationSearchParams = createSerializer(searchParams)

const emptyPaginationStateParams = {
    [paginationSearchParamKeys.before]: null,
    [paginationSearchParamKeys.after]: null,
    [paginationSearchParamKeys.first]: null,
    [paginationSearchParamKeys.last]: null,
};

return { setPaginationState, serializePaginationSearchParams, emptyPaginationStateParams }
```

```ts
// callsite

// How should I type `emptyPaginationStateParams` such that this works?
setPaginationState(emptyPaginationStateParams, { history: "replace" })
serializePaginationSearchParams(emptyPaginationStateParams)
```

**Comment by franky47** [comment, marked as answer]
For the clearing part, you should be able to call `setPaginationState(null, { /* options here */ })`. This will clear all search params that were declared to the `useQueryStates` hook, without having to enumerate them _(available since 1.19.1)_.

That being said, the serializer doesn't have this behaviour, but it could be something nice to add. Would you like to open a PR?

Something else that might be of use is the [`inferParserType`](https://nuqs.47ng.com/docs/utilities#parser-type-inference) utility.

**Reply by neefrehman** [reply]
Thanks @franky47! I've just opened a PR here: https://github.com/47ng/nuqs/pull/657
